### PR TITLE
WTB-441 - returning validation reason

### PR DESCRIPTION
**Pull request by cemusta**
- validation part extracted into middlewares ( moved from controllers to separate file )
 - validator is returning a human-readable validation error

### CODE REVIEW

File: src/api/middleware/validator/entityValidator.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  return (request, _res, next) => {
+    const input = plainToClass(type, request.body, {
+      excludeExtraneousValues: true,
+    });
+
+    const errors = validateSync(input, { skipMissingProperties: true });
+    if (errors.length > 0) {
+      return next(new BadRequestError(errors.toString()));
+    }
+    if (Object.values(input).every((x) => x == null)) {
```
Comment (darrylhebbes) [inline src/api/middleware/validator/entityValidator.ts:17]: So we allow empty string values ?

Comment (cemusta) [reply src/api/middleware/validator/entityValidator.ts:17]: Nice catch :)

Comment (cemusta) [reply src/api/middleware/validator/entityValidator.ts:17]: We were but I've added  `@NotEquals('')` to request models. So its catched above.

What I've wanted to catch here is if the input is completely empty or not, when all properties can have IsOptional

Comment (darrylhebbes) [reply src/api/middleware/validator/entityValidator.ts:17]: Ok cool, and there is not an @NotNull decorator?

Comment (darrylhebbes) [reply src/api/middleware/validator/entityValidator.ts:17]: or `@NotEquals(null)`

File: src/api/model/RequestedDocumentPatch.ts [hunk 2]
```diff
@@ -1,5 +1,5 @@
 import { Expose } from 'class-transformer';
-import { IsBoolean, IsOptional, IsString } from 'class-validator';
+import { IsBoolean, IsOptional, IsString, NotEquals } from 'class-validator';
```
Comment (darrylhebbes) [inline src/api/model/RequestedDocumentPatch.ts:2]: perhaps `@IsNotEmpty()` is a better choice https://github.com/typestack/class-validator#validation-decorators

Comment (darrylhebbes) [reply src/api/model/RequestedDocumentPatch.ts:2]: checks: "Checks if given value is not empty (!== '', !== null, !== undefined)."

Comment (darrylhebbes) [reply src/api/model/RequestedDocumentPatch.ts:2]: or its negation `@IsEmpty()`

File: src/api/middleware/validator/entityValidator.ts (hunk 1 above)
Comment (cemusta) [reply src/api/middleware/validator/entityValidator.ts:17]: NotNull kind of validator is a compiting validator with isOptional, will cause validation to cause all the time. I'm kinda new with this class-validators so I learned this recently.

File: src/api/model/RequestedDocumentPatch.ts (hunk 2 above)
Comment (cemusta) [reply src/api/model/RequestedDocumentPatch.ts:2]: I guess its not directly working with `@IsOptional` will check

Comment (cemusta) [reply src/api/model/RequestedDocumentPatch.ts]: it worked, however that part is currently not used (text edit part I mean) we can add max min lenght validators as well in future.

**Review by darrylhebbes — APPROVED** [review]

**Pull request merged**